### PR TITLE
feat: CSV data load improvements - auto-mapping, validation, analyze mode

### LINKED ISSUE #125: data load: improve auto-mapping UX to prevent silent empty records
## Problem

When using `ppds data load` with a CSV file, if column names don't exactly match Dataverse attribute names, the auto-mapping silently skips them with only a warning. The load "succeeds" but records are created with empty fields.

**Example:**
```
CSV columns:       code, city, state, county, lat, lon
Dataverse attrs:   ppds_code, ppds_city, ppds_stateid, ppds_county, ppds_latitude, ppds_longitude
```

Result: 42K records created successfully, but ALL fields are empty because nothing matched.

## Personas Affected

| Persona | Impact |
|---------|--------|
| **Extension** | Silent failures break automation |
| **Users** | No indication data would be empty until they check Dataverse |
| **Claude/AI** | Can't self-correct without structured feedback |

### P0: Fail-Safe Threshold
If <50% columns match, **refuse to proceed** by default:
```
Error: Only 1/7 columns matched (14%)
  ✓ id → ppds_stateid
  ✗ code → (no match) - did you mean: ppds_code?
  ✗ city → (no match) - did you mean: ppds_cityid?
  ...

This would create records with mostly empty fields.
Use --force to proceed, or --generate-mapping to fix.
```

### P1: Prefix-Aware Auto-Mapping
Try matching with common patterns:
- `abbreviation` → `ppds_abbreviation` (prefix match)
- `Name` → `ppds_name` (case-insensitive)
- `state` → `ppds_stateid` (lookup suffix awareness)

Could be opt-in (`--auto-prefix ppds_`) or smart by default.

### P2: Enhanced `--analyze` Mode
```
$ ppds data load -e ppds_state -f states.csv --analyze

=== Mapping Analysis ===
Entity: ppds_state | CSV: 51 rows

Column Mappings:
  CSV Column      Dataverse Attr         Status
  ───────────     ──────────────         ──────
  abbreviation    -                      ✗ No match
                  → ppds_abbreviation    (95% similar)
  name            -                      ✗ No match  
                  → ppds_name            (100% similar)

Recommendation: Use --generate-mapping mapping.json
```

### P3: JSON Output for Extension/Claude
```json
{
  "analysis": {
    "matched": [{"csv": "id", "attr": "ppds_stateid"}],
    "unmatched": [
      {"csv": "abbreviation", "suggestions": ["ppds_abbreviation"]}
    ],
    "matchRate": 0.33,
    "wouldSucceed": false,
    "recommendation": "generate-mapping"
  }
}
```

## Discovered In
Testing `ppds data load` with geo-data CSV (42K ZIP codes) in ppds-demo.

## Related
- `scripts/Test-PpdsDataLoad.ps1` in ppds-demo (blocked by this issue)

### LINKED ISSUE #126: data load: alternate keys with Lookup fields fail to resolve
## Problem

When using `ppds data load` with a composite alternate key that includes a Lookup field, the lookup value isn't resolved before being set in KeyAttributes.

**Example:**
```bash
ppds data load -e ppds_city -f cities.csv -k ppds_name,ppds_stateid -m cities.json
```

Where `ppds_stateid` is a Lookup to `ppds_state`, and the CSV contains `AK` (state abbreviation).

**Error:**
```
TYPE_COERCION_FAILED: Cannot convert key value 'AK' to Lookup
```

## Root Cause

In `CsvDataLoader.cs` lines 354-382, when processing alternate key fields:

1. `CoerceValue` is called for each key field value
2. For Lookup types, `CoerceValue` returns `null` (expecting LookupResolver to handle)
3. The key processing code interprets `null` as a failure
4. The lookup is never resolved to an EntityReference

## Expected Behavior

Alternate key fields that are Lookups should:
1. Use LookupResolver to resolve the value (e.g., 'AK' → EntityReference to ppds_state)
2. Set the resolved EntityReference in KeyAttributes

## Workaround

None - composite alternate keys with Lookup fields cannot be used until this is fixed.

## Related

- #125 (auto-mapping UX)
- Discovered testing geo-data load in ppds-demo

### LINKED ISSUE #134: test: add unit tests for CSV data loader
## Summary

Add unit tests for CSV data loader classes. Currently only 2 tests exist for 6 source files.

**Priority:** P3 (newer feature, expand existing coverage)

## Background

The CSV loader (`ppds data load`) imports CSV files into Dataverse. It handles:
- CSV parsing and mapping
- Lookup resolution (resolving references to related records)
- Auto-mapping generation
- Record transformation

Some tests exist but coverage is thin.

### CSV Loader (`src/PPDS.Cli/CsvLoader/`)
- [ ] `CsvDataLoader.cs` - main loader orchestration
- [ ] `LookupResolver.cs` - relationship resolution
- [ ] `MappingGenerator.cs` - auto-mapping from CSV headers
- [ ] `CsvRecordMapper.cs` - individual record mapping
- [ ] `CsvImportResult.cs` - result aggregation

### CsvDataLoader
- Parse valid CSV with headers
- Handle missing required columns
- Handle type mismatches
- Progress callback invocation

### LookupResolver
- Resolve by primary key
- Resolve by alternate key
- Handle unresolved lookups
- Cache hit/miss behavior

### MappingGenerator
- Auto-map matching column names
- Handle case-insensitive matching
- Skip unmappable columns
- Generate mapping report

### CsvRecordMapper
- Map simple types (string, int, decimal)
- Map date/datetime
- Map option sets
- Map lookups

## Acceptance Criteria

- [ ] Each file has test coverage
- [ ] Edge cases covered (empty CSV, malformed data)
- [ ] Lookup resolution tested with mocks
- [ ] Mapping generation tested
- [ ] All tests pass in CI

## Related

- Parent epic: `epic:testing`
- Feature: #36 (CSV import command)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by joshsmithxrm**
## Summary

This PR implements comprehensive improvements to the CSV data loading functionality:

- **#126 Fix**: Resolve lookup fields in alternate keys - composite keys with lookup fields now work correctly
- **#125 Features**: 
  - Require 100% column match or mapping file (exit code 7 if incomplete)
  - Validate mapping file completeness (exit code 8 if invalid)
  - Prefix-aware auto-mapping (`name` → `ppds_name`)
  - Self-documenting mapping files with `_note`, `_status`, `_csvSample` hints
  - `--analyze` mode for mapping preview (text and JSON output)
  - `--force` option to bypass incomplete auto-mapping
  - Mapping file schema version validation
- **#134 Tests**: Unit tests for all new CSV loader types

### New Exit Codes
| Code | Meaning |
|------|---------|
| 0 | Success |
| 7 | Mapping required (auto-mapping incomplete) |
| 8 | Validation error (mapping file invalid) |

### New CLI Options
```bash
ppds data load -e ppds_state -f states.csv --analyze          # Preview mapping
ppds data load -e ppds_state -f states.csv --analyze -o json  # JSON output
ppds data load -e ppds_state -f states.csv --force            # Skip unmatched columns
ppds data load -e ppds_state -f states.csv --generate-mapping # Create mapping file
```

## Test plan

- [x] Build succeeds with 0 warnings
- [x] All unit tests pass (541 CLI tests)
- [x] CLI smoke tests pass (`--analyze`, `--force`, prefix matching, exit codes)
- [x] Composite key with lookup field resolves correctly
- [x] Demo scripts updated and tested (see ppds-demo#65)

## Related

- Closes #125
- Closes #126
- Closes #134
- Follow-up: #135 (update/delete commands)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by joshsmithxrm** [comment]
## Bot Review Feedback - Addressed

Thanks to @gemini-code-assist[bot] and @copilot-pull-request-reviewer[bot] for the thorough reviews!

### Fixed in commit 46dbea7

**Code Duplication (7 instances)**
- Extracted shared matching logic to `ColumnMatcher.cs` - single source of truth for:
  - `ExtractPublisherPrefix`
  - `NormalizeForMatching`
  - `FindSimilarAttributes`
  - `IsLookupAttribute`
  - `IsOptionSetAttribute`
  - `BuildAttributeLookup`
  - `MaxSampleValues` constant
- Both `CsvDataLoader` and `MappingGenerator` now use `ColumnMatcher.*` methods

**Schema Version Scattered (3 locations)**
- Centralized in `CsvMappingSchema.cs` with:
  - `CurrentVersion` constant
  - `SchemaUrl` constant
  - `ValidateVersion()` method
  - `ParseVersion()` method
- `LoadCommand` and `MappingGenerator` now reference single source of truth

**Duplicate ExitCodes Classes**
- Deleted `src/PPDS.Cli/Commands/ExitCodes.cs`
- All code now uses `PPDS.Cli.Infrastructure.Errors.ExitCodes`

### New Tests
- `ColumnMatcherTests.cs` - 24 tests
- `CsvMappingSchemaTests.cs` - 18 tests

### Documentation
- Created ADR-0009: CSV Mapping Schema and Versioning
- Updated CHANGELOG.md

### Future Work
- Created issue #147 for JSON data export format consideration

**Comment by joshsmithxrm** [comment]
@gemini-code-assist review

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 1]
```diff
@@ -18,22 +18,198 @@ namespace PPDS.Cli.CsvLoader;
...
+            Columns = columns,
+            Recommendations = recommendations
+        };
+    }
+
+    private async Task<(string[] Headers, Dictionary<string, List<string>> SampleValues)> ReadCsvHeadersAndSamplesAsync(
+        string csvPath,
+        CancellationToken cancellationToken)
+    {
+        const int maxSampleValues = 3;
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs]: Fixed in ad3d356 - now uses ColumnMatcher.MaxSampleValues

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 2]
```diff
@@ -18,22 +18,198 @@ namespace PPDS.Cli.CsvLoader;
...
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs]: Fixed in ad3d356 - now uses ColumnMatcher.MaxSampleValues

[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs]: Fixed in ad3d356 - now uses ColumnMatcher.MaxSampleValues

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 3]
```diff
@@ -18,22 +18,198 @@ namespace PPDS.Cli.CsvLoader;
...
+        }
+
+        if (lookupCount > 0)
+        {
+            recommendations.Add($"{lookupCount} lookup field(s) detected. Review generated mapping to configure resolution.");
+        }
+
+        if (unmatchedCount == 0 && lookupCount == 0)
+        {
+            recommendations.Add("All columns matched. Ready to load with: ppds data load --entity " + entityLogicalName + " --file " + Path.GetFileName(csvPath));
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs]: Fixed in ad3d356 - using string interpolation

File: src/PPDS.Cli/Commands/Data/LoadCommand.cs [hunk 4]
```diff
@@ -365,6 +441,108 @@ private static async Task<int> GenerateMappingAsync(
...
+    private static async Task<int> AnalyzeMappingAsync(
+        IDataverseConnectionPool pool,
+        string entityName,
+        string csvPath,
+        OutputFormat outputFormat,
+        CancellationToken cancellationToken)
+    {
+        Console.Error.WriteLine($"Analyzing mapping for '{entityName}'...");
+
+        var loader = new CsvDataLoader(pool, null, null);
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/Commands/Data/LoadCommand.cs]: Fixed in ad3d356 - logger now passed to AnalyzeMappingAsync

File: docs/adr/0009_CSV_MAPPING_SCHEMA.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+- Users can add notes without affecting behavior
+- Forward compatibility - unknown properties are ignored
+
+### Forward Compatibility
+
+The `CsvMappingConfig` class uses `[JsonExtensionData]` to capture unknown properties:
+
+```csharp
+[JsonExtensionData]
+public Dictionary<string, JsonElement>? Extensions { get; set; }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply docs/adr/0009_CSV_MAPPING_SCHEMA.md]: Fixed in ad3d356 - documentation now shows ExtensionData to match actual property

File: src/PPDS.Cli/CsvLoader/SchemaVersionException.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    /// The current CLI version.
+    /// </summary>
+    public string CliVersion { get; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SchemaVersionException"/> class.
+    /// </summary>
+    public SchemaVersionException(string fileVersion, string cliVersion)
+        : base($"Mapping file version {fileVersion} is not compatible with CLI version {cliVersion}. " +
+               $"Please upgrade to CLI v{GetMajorVersion(fileVersion)}.x or regenerate the mapping file.")
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/SchemaVersionException.cs]: Fixed in ad3d356 - message now just says 'regenerate the mapping file using this CLI version'

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 7]
```diff
@@ -18,22 +18,198 @@ namespace PPDS.Cli.CsvLoader;
...
+        // No match
+        var suggestions = FindSimilarAttributes(header, attributesByName, prefix);
+        return new ColumnAnalysis
+        {
+            CsvColumn = header,
+            IsMatched = false,
+            Suggestions = suggestions.Count > 0 ? suggestions : null,
+            SampleValues = samples?.Count > 0 ? samples : null
+        };
+    }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:192]: Already addressed - column matching logic was extracted to ColumnMatcher class in 513a027

File: src/PPDS.Cli/CsvLoader/MappingGenerator.cs [hunk 8]
```diff
@@ -173,20 +173,54 @@ private ColumnMappingEntry CreateMappingEntry(
...
+    {
+        // Standard Dataverse entity naming: <publisher>_<name>
+        // Extract the prefix including underscore: "ppds_city" → "ppds_"
+        var underscoreIndex = entityLogicalName.IndexOf('_');
+        if (underscoreIndex > 0)
+        {
+            return entityLogicalName[..(underscoreIndex + 1)];
+        }
+        return null;
     }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/MappingGenerator.cs]: Already addressed - shared logic (ExtractPublisherPrefix, FindSimilarAttributes, matching) was extracted to ColumnMatcher class in 513a027

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 9]
```diff
@@ -223,20 +448,95 @@ private async Task<Dictionary<string, ColumnMappingEntry>> AutoMapColumnsAsync(
...
+    {
+        // Standard Dataverse entity naming: <publisher>_<name>
+        // Extract the prefix including underscore: "ppds_city" → "ppds_"
+        var underscoreIndex = entityLogicalName.IndexOf('_');
+        if (underscoreIndex > 0)
+        {
+            return entityLogicalName[..(underscoreIndex + 1)];
+        }
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs]: Already addressed - ExtractPublisherPrefix is in ColumnMatcher (513a027)

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 10]
```diff
@@ -223,20 +448,95 @@ private async Task<Dictionary<string, ColumnMappingEntry>> AutoMapColumnsAsync(
...
+                results.Add((attrName, score, false));
+            }
+        }
+
+        return results
+            .OrderByDescending(r => r.PrefixMatch) // Prefix matches first
+            .ThenBy(r => r.Score)
+            .Take(3)
+            .Select(r => r.Name)
+            .ToList();
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs]: Already addressed - FindSimilarAttributes is in ColumnMatcher (513a027)

File: src/PPDS.Cli/Commands/Data/LoadCommand.cs [hunk 11]
```diff
@@ -469,4 +642,157 @@ private static void WriteJsonResult(LoadResult result)
...
+                missingMappings = ex.MissingMappings,
+                staleMappings = ex.StaleMappings,
+                suggestion = "Update the mapping file to configure all columns (set 'field' or 'skip: true')"
+            }
+        };
+
+        Console.WriteLine(JsonSerializer.Serialize(output, JsonOptions));
+    }
+
+    private const string CurrentSchemaVersion = "1.0";
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/Commands/Data/LoadCommand.cs]: Already addressed - version constants are in CsvMappingSchema class (513a027)

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:192]: Already addressed - matching logic extracted to ColumnMatcher class (513a027)

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 12]
```diff
@@ -18,22 +18,198 @@ namespace PPDS.Cli.CsvLoader;
...
+                    && sampleValues[header].Count < maxSampleValues
+                    && !sampleValues[header].Contains(value))
+                {
+                    sampleValues[header].Add(value);
+                }
+            }
+            rowCount++;
+        }
+
+        return (headers, sampleValues);
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs]: Declining - MappingGenerator now uses ColumnMatcher.MaxSampleValues. The remaining duplication is acceptable since both methods have different CSV configurations.

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 13]
```diff
@@ -395,6 +735,12 @@ private static string NormalizeForMatching(string value)
                     continue;
                 }
 
+                // Skip fields already processed as alternate key attributes
+                if (processedKeyFields.Contains(fieldName))
+                {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:649]: Declining - processedKeyFields uses StringComparer.OrdinalIgnoreCase (line 546), and mapping.Field is the actual attribute name. The casing is handled correctly by the comparer.

File: src/PPDS.Cli/CsvLoader/MappingGenerator.cs [hunk 14]
```diff
@@ -247,15 +281,16 @@ private ColumnMappingEntry CreateMatchedEntry(
...
+        var similarAttributes = FindSimilarAttributes(header, entityMetadata, prefix);
 
         return new ColumnMappingEntry
         {
-            Skip = true,
-            Status = "no-match",
-            Note = "No matching attribute found. Set 'field' to map, or keep 'skip' to ignore.",
+            Field = null, // Forces explicit decision
+            Status = "needs-configuration",
+            Note = "No matching attribute found. Set 'field' to map to an attribute, or set 'skip: true' to ignore this column.",
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/MappingGenerator.cs:237]: Declining - intentional design. Field=null forces explicit user decision; validation uses IsNullOrEmpty to catch both null and empty string cases.

File: tests/PPDS.Cli.Tests/CsvLoader/MappingIncompleteExceptionTests.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    [Fact]
+    public void EmptyUnmatchedColumns_IsValid()
+    {
+        var exception = new MappingIncompleteException(5, 5, new List<UnmatchedColumn>());
+
+        Assert.Empty(exception.UnmatchedColumns);
+        Assert.Equal(5, exception.MatchedColumns);
+    }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply tests/PPDS.Cli.Tests/CsvLoader/MappingIncompleteExceptionTests.cs:58]: Declining - test validates constructor behavior accepts empty lists. The caller decides when to throw the exception, not the constructor.

File: src/PPDS.Cli/Commands/ExitCodes.cs [hunk 16]
```diff
@@ -16,4 +16,19 @@ public static class ExitCodes
...
+    public const int AuthError = 5;
+
+    /// <summary>Resource not found (profile, environment, file).</summary>
+    public const int NotFoundError = 6;
+
+    /// <summary>Mapping required - auto-mapping incomplete, need --generate-mapping or --force.</summary>
+    public const int MappingRequired = 7;
+
+    /// <summary>Validation error - incomplete mapping file, schema mismatch, etc.</summary>
+    public const int ValidationError = 8;
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/Commands/ExitCodes.cs]: False positive - file src/PPDS.Cli/Commands/ExitCodes.cs does not exist. Only src/PPDS.Cli/Infrastructure/Errors/ExitCodes.cs exists.

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 17]
```diff
@@ -18,22 +18,198 @@ namespace PPDS.Cli.CsvLoader;
...
+    /// <param name="logger">Optional logger.</param>
     public CsvDataLoader(
         IDataverseConnectionPool pool,
-        IBulkOperationExecutor bulkExecutor,
+        IBulkOperationExecutor? bulkExecutor,
         ILogger<CsvDataLoader>? logger = null)
     {
         _pool = pool ?? throw new ArgumentNullException(nameof(pool));
-        _bulkExecutor = bulkExecutor ?? throw new ArgumentNullException(nameof(bulkExecutor));
+        _bulkExecutor = bulkExecutor;
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:36]: Declining - documented design. Constructor docs explicitly state: 'Required for LoadAsync, optional for AnalyzeAsync'. The InvalidOperationException in LoadAsync is intentional.

File: tests/PPDS.Cli.Tests/CsvLoader/MappingValidationExceptionTests.cs [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        var exception = new MappingValidationException(
+            new List<string>(),
+            new List<string>(),
+            new List<string>());
+
+        Assert.Empty(exception.UnconfiguredColumns);
+        Assert.Empty(exception.MissingMappings);
+        Assert.Empty(exception.StaleMappings);
+    }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply tests/PPDS.Cli.Tests/CsvLoader/MappingValidationExceptionTests.cs:84]: Declining - test validates constructor accepts empty lists. In real scenarios, this exception won't be thrown when all lists are empty (caller logic prevents it).

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 19]
```diff
@@ -489,10 +835,92 @@ private static string NormalizeForMatching(string value)
...
+        }
+
+        // Check for CSV columns not in mapping
+        foreach (var csvColumn in csvColumns)
+        {
+            if (!mappingColumns.Contains(csvColumn))
+            {
+                missingMappings.Add(csvColumn);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:787]: Declining - style preference. foreach+if is readable and performs equally; using .Where() for filtering a foreach is not a requirement.

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 20]
```diff
@@ -489,10 +835,92 @@ private static string NormalizeForMatching(string value)
...
+
+        // Check for mapping columns not in CSV (stale entries - warning only)
+        foreach (var mappingColumn in mappingColumns)
+        {
+            if (!csvColumns.Contains(mappingColumn))
+            {
+                staleMappings.Add(mappingColumn);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:796]: Declining - style preference. foreach+if is readable and performs equally; using .Where() for filtering a foreach is not a requirement.

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 21]
```diff
@@ -54,11 +236,48 @@ public async Task<LoadResult> LoadAsync(
             options.EntityLogicalName, cancellationToken);
 
         // 2. Build attribute lookup
-        var attributesByName = BuildAttributeLookup(entityMetadata);
+        var attributesByName = ColumnMatcher.BuildAttributeLookup(entityMetadata);
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:239]: Declining - AnalyzeAsync shows all mappable attributes for the analysis preview. Filtering to valid-for-update is done in the loading path at runtime.

File: src/PPDS.Cli/CHANGELOG.md [hunk 22]
```diff
@@ -44,6 +44,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - **Progress output to stderr** - Progress messages now write to stderr, enabling clean piping: `ppds data export -f json | jq` ([#76](https://github.com/joshsmithxrm/ppds-sdk/issues/76))
 - **Status messages to stderr** - Operational messages (connecting, authenticating, etc.) now write to stderr, keeping stdout for data only ([#76](https://github.com/joshsmithxrm/ppds-sdk/issues/76))
+- **Centralized CSV column matching** - Extracted shared column-to-attribute matching logic to `ColumnMatcher` class for code reuse between data loading and mapping generation
+- **Centralized schema versioning** - Schema version constants centralized in `CsvMappingSchema` class (single source of truth)
+- **ADR-0009** - Architecture decision record documenting CSV mapping schema versioning and underscore prefix convention
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CHANGELOG.md:49]: Declining - changes are correctly under the [Unreleased] section (line 8). This is the standard changelog format.

File: src/PPDS.Cli/CsvLoader/ColumnMatcher.cs [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+        out AttributeMetadata? found)
+    {
+        foreach (var kvp in attributes)
+        {
+            if (NormalizeForMatching(kvp.Key) == normalizedHeader)
+            {
+                found = kvp.Value;
+                return true;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/ColumnMatcher.cs:236]: Declining - style preference. foreach+if is readable and performs equally.

File: src/PPDS.Cli/Commands/Data/LoadCommand.cs [hunk 24]
```diff
@@ -469,4 +642,157 @@ private static void WriteJsonResult(LoadResult result)
...
+        }
+
+        // Higher minor version → warn
+        if (fileParts.Minor > cliParts.Minor)
+        {
+            if (outputFormat != OutputFormat.Json)
+            {
+                Console.Error.WriteLine($"Warning: Mapping file version {fileVersion} is newer than CLI version {CurrentSchemaVersion}. " +
+                    "Some features may be ignored.");
+            }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/Commands/Data/LoadCommand.cs]: False positive - the suggested code references variables (fileParts, CurrentSchemaVersion) that don't exist in LoadCommand.cs. The version check is in CsvMappingSchema.ValidateVersion() and is already properly combined.

File: src/PPDS.Cli/Commands/Data/LoadCommand.cs [hunk 25]
```diff
@@ -469,4 +642,157 @@
...
+        }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/Commands/Data/LoadCommand.cs]: Duplicate of Copilot comment #2659875715 - false positive, see reply there

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 26]
```diff
@@ -489,10 +835,92 @@
...
+        }
+
+        // Check for CSV columns not in mapping
+        foreach (var csvColumn in csvColumns)
+        {
+            if (!mappingColumns.Contains(csvColumn))
+            {
+                missingMappings.Add(csvColumn);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:787]: Duplicate of Copilot comment #2659875711 - declining (style preference)

File: src/PPDS.Cli/CsvLoader/CsvDataLoader.cs [hunk 27]
```diff
@@ -489,10 +835,92 @@
...
+
+        // Check for mapping columns not in CSV (stale entries - warning only)
+        foreach (var mappingColumn in mappingColumns)
+        {
+            if (!csvColumns.Contains(mappingColumn))
+            {
+                staleMappings.Add(mappingColumn);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/CsvDataLoader.cs:796]: Duplicate of Copilot comment #2659875713 - declining (style preference)

File: src/PPDS.Cli/CsvLoader/ColumnMatcher.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply src/PPDS.Cli/CsvLoader/ColumnMatcher.cs:236]: Duplicate of Copilot comment #2659915164 - declining (style preference)
